### ISSUE TITLE
Pango-CRITICAL pango_font_describe pango_font_description_get_variant 

### ISSUE THREAD

**Issue by mfg92** [darktable-org/darktable#10882]
**Describe the bug/issue**

If I start darktable on windows using bash (aka git-bash, this is part of git for windows https://gitforwindows.org/) the console is spamed hundreds of times with these two lines:
```
(darktable.exe:2996): Pango-CRITICAL **: 19:09:04.794: pango_font_description_get_variant: assertion 'desc != NULL' failed
(darktable.exe:2996): Pango-CRITICAL **: 19:09:04.794: pango_font_describe: assertion 'font != NULL' failed
```

Why do I use bash to start darktable? Because I want to inspect some performance issues using `-d opencl -d perf
`.

**To Reproduce**
Start darktable using bash without any arguments (CMD and powershell do not cause this issue).

**Expected behavior**
No _Pango-CRITICAL_ error messages in the console.

**Platform**
_Please fill as much information as possible in the list given below. Please state "unknown" where you do not know the answer and remove any sections that are not applicable _

* darktable version: 3.8.0 (official build)
* OS                :  Win10 Pro 21H2
* Memory            :  32GB
* Graphics card     :  Radeon RX 480 8GB
* OpenCL activated  : yes

**Comment by Nilvus** [comment]
Seems really similar to that issue #10588 and so more a Pango issue than darktable one. I don't use Windows but I can't reproduce on Linux I'm using and regarding comments in #10588, you probably have Pango 1.50 which is known to be bugged!

**Comment by gi-man** [comment]
@Nilvus I'm not sure I agree this is a duplicate from $10558. There is no crash and everything seems to work correctly. 

I'm on Windows and I'm seeing the same error. It is a minor issue. Mostly annoying when you look at the log after -d perf 

Is assertion a font type?

**Comment by Nilvus** [comment]
> @Nilvus I'm not sure I agree this is a duplicate from $10558. There is no crash and everything seems to work correctly.
> 
> I'm on Windows and I'm seeing the same error. It is a minor issue. Mostly annoying when you look at the log after -d perf
> 
> Is assertion a font type?

It's not a complete duplicate I agree but it could anyway be for same origin. I don't have those lines on Linux and I don't have Pango 1.50 too. And Pango 1.50 is known to have issues such like those lines. So please check which Pango version you have. It's probably not a darktable issue.

**Comment by mfg92** [comment]
Thank you @Nilvus for your investigation. 
I do not know which pango version darktable is using. I'm using the official darktable release and I assume that pango comes as part of the installations exe file. I found these dll files in the darktable installation folder:

- libpango-1.0-0.dll
- libpangocairo-1.0-0.dll
- libpangoft2-1.0-0.dll
- libpangowin32-1.0-0.dll

Then I opened _libpango-1.0-0.dll_ in VS Code and found the string _pango-1.50.0_ multiple times, so i guess you are right and my darktable installation is using this problematic release of pango.

In one of the [comments ](https://github.com/darktable-org/darktable/issues/10588#issuecomment-989162787) of #10588 @paolodepetrillo pointed out that there is already a fix but no release yet. This seems to have changed: https://github.com/GNOME/pango/blob/main/NEWS

**Comment by kmilos** [comment]
> so i guess you are right and my darktable installation is using this problematic release of pango

It is not. The 1.50 version dt was built against for the 3.8 release includes the special patch for the crash.

This issue is completely different and also happens w/ pango 1.52

**Comment by Nilvus** [comment]
> > so i guess you are right and my darktable installation is using this problematic release of pango
> 
> It is not. The 1.50 version dt was built against for the 3.8 release includes the special patch for the crash.
> 
> This issue is completely different and also happens w/ pango 1.50.3

Thanks @kmilos for your precision. That's so a different one and probably a specific Windows one. I can't reproduce on my side on Linux.

**Comment by gi-man** [comment]
@Nilvus is assertion a font (eg like Arial)? Is part of your Linux system? 

I found a font named assertion online, but it is not part of windows.

**Comment by gi-man** [comment]
Nah. That's not it. Nevemind

**Comment by kmilos** [comment]
[It is not a font.](https://www.cplusplus.com/reference/cassert/assert/)

**Comment by ChristianBirzer** [comment]
I also get the pango font assertions and it is very annoying when debugging dt because the log is just flooded with these messages.
I just encountered that the assertions are only printed when I use the css styles "darktable" or "darktable icons" (which I did prefer because the letterers are a bigger 😃)
The others ("elegant") styles all have explicit font declarations in the css whereas the darktable.css and darktable-icons.css do not. I guess this is the reason for the assertions.
@mfg92  could you please check if you can get rid off the messages when switching the design in preferences? If so, I'm sure we could find a bigger font that works on windows too - Segoe UI (without light or condensed) looks very good at first glance.

**Comment by gi-man** [comment]
@ChristianBirzer I just tested with the Elegant style and the pango assertions do not show up. If I return it to "darktable", it comes back.

**Comment by mfg92** [comment]
@ChristianBirzer you are right, when I change to any of the _darktable elegant *_ themes I do not get those messages. When I switch back to any of the other themes I get the error messages back.

**Comment by ChristianBirzer** [comment]
Thanks a lot. When I find some time during the next week I'll contribute an update for these themes.

**Comment by Nilvus** [comment]
Just in case, someone don't read my answer in @ChristianBirzer, I copy/paste it here:

it's not good. @ChristianBirzer, what you do is implement partly darker elegant theme in darktable main theme. And darktable main theme allow to use OS font used. So, your issue could be solved by only 2 ways (and absolutely not by changing that theme):

    use darktable-elegant-darker theme. It's just darktable theme with Roboto font, and was done to avoid issue you point here (not only but was a reason)
    change font used by your OS

This is so not possible to change that and would probably made some users scream! Some don't like Roboto fonts!

**Comment by Nilvus** [comment]
Those warnings are probably here depending on which font used. So, if you set another font, you will probably see the warnings disappear and make those visible again with another font. So, if this is confirm, it's probably more a Pango issue than something to fix in darktable.

**Comment by ChristianBirzer** [comment]
> So, if you set another font, you will probably see the warnings disappear and make those visible again with another font.

Question is, how to set an other font. The warnings occur on windows (at least). On Windows it is not common to set an other system font, nor a special font for one application.

So maybe using an other css that explicitly sets a font to avoid this problem isn't that bad, at least on Windows. I agree that overriding the system font isn't a good idea if that was the intention of that default theme but it seems to do no good job in an Windows environment.
Maybe add an other theme that looks good at win for win users? I don't know...

BTW: The elegant themes use light fonts that are hard to read for me.

**Comment by Nilvus** [comment]
> > So, if you set another font, you will probably see the warnings disappear and make those visible again with another font.
> 
> Question is, how to set an other font. The warnings occur on windows (at least). On Windows it is not common to set an other system font, nor a special font for one application.

It's possible on any main OS. You should find that on Windows preferences. I don't know exactly where as I don't use Windows. Of course, it's not common and I would be the first to like seeing the fact that main darktable theme use OS font disappear. That had been tried and found quite good opposition. I agree that it is not common to set another font for one application in the OS. But that not what we said. The main theme is done to use, by default OS font not to adapt OS font to the application. It's 2 different things. Again, I think you confuse symptom and cause. The problem is quite probably not the font, nor the OS, about that issue, it's certainly a bug in Pango library for some specific font. On my side, with default OS font (I use Linux Debian with Gnome 41 desktop environment), I don't have any issue nor any line about Pango (or something else) when I launch darktable with darktable main theme like you.

> So maybe using an other css that explicitly sets a font to avoid this problem isn't that bad, at least on Windows. I agree that overriding the system font isn't a good idea if that was the intention of that default theme but it seems to do no good job in an Windows environment. Maybe add an other theme that looks good at win for win users? I don't know...
> BTW: The elegant themes use light fonts that are hard to read for me.

You could do that for you by setting font by just copy/paste all darktable-elegant-darker.css file to CSS editor in general tab of main darktable preferences and change font to those you want.

Doing a theme for an OS is a workaround not a fix. And so not a good way. And again, even if it could be good to fix that, it's not a reallly critical issue, so to fix it, be sure that the fix is really acceptable on all sides and avoid other not wanted side-effect.

It could be good too to post an issue in Pango Github page (and search before if there's already similar one posted for darktable or another one).

**Comment by Nilvus** [comment]
> BTW: The elegant themes use light fonts that are hard to read for me.

And do you have all the Roboto fonts defined in the darktable-elegant-darkter.css file installed?

**Comment by mfg92** [comment]
I never used the elegant themes because the fonts were hard to read like @ChristianBirzer mentioned.
Now I have installed the Roboto font family from [here](https://fonts.google.com/specimen/Roboto) and it looks way better now. It is better to read but still _darktable-icons_ is the best in this regard:
darktable-elegant-darker:
![darktable-elegant-darker](https://user-images.githubusercontent.com/11365250/151251250-31d75ff2-0f45-4a91-b22d-f08061e09299.PNG)
darktable-icons:
![darktable-icons](https://user-images.githubusercontent.com/11365250/151251259-de3c9358-1f03-4b8e-a486-8d0e5dcd1a64.PNG)

**Comment by Nilvus** [comment]
> Now I have installed the Roboto font family from [here](https://fonts.google.com/specimen/Roboto) and it looks way better now. It is better to read but still _darktable-icons_ is the best in this regard:

Yes, it's always better to have fonts used by a theme to display it correctly. About the fact that it looks way better for is just a matter of taste. And we all have different taste. I found Roboto theme way better (and I know I'm not the only one) and there's probably quite a lot of people on both sides (prefer default themes so with default OS fonts or elegant themes). That's why we have different themes and main one use OS font and other based on Roboto (and so that was not a good way to change main theme to use Roboto font). Last thing, it's also the reason why darktable have a CSS editor inside it's main preferences (and in general theme).

**Comment by kmilos** [comment]
[This is now causing a crash on Windows](https://github.com/darktable-org/darktable/issues/11407) (due to harfbuzz API changes), and has been reported upstream, so please help over there if you can: https://gitlab.gnome.org/GNOME/pango/-/issues/680

**Comment by kmilos** [comment]
This can be closed now, fixed by [patched pango-1.50.6-3 that is now available in the MSYS2 repo](https://github.com/msys2/MINGW-packages/commit/99232723b6d7b10517ca15eacdc9d10cff441341).

**Comment by TurboGit** [comment]
Closing.
